File: src/configpaths.cpp
#include "configpaths.h"

namespace Newsboat {

ConfigPaths::ConfigPaths()
	: rs_object(configpaths::bridged::create())
{
}

bool ConfigPaths::initialized() const
{
	return Newsboat::configpaths::bridged::initialized(*rs_object);
}

std::string ConfigPaths::error_message() const
{
	return std::string(Newsboat::configpaths::bridged::error_message(*rs_object));
}

void ConfigPaths::process_args(const CliArgsParser& args)
{
	Newsboat::configpaths::bridged::process_args(*rs_object, args.get_rust_ref());
}

bool ConfigPaths::try_migrate_from_newsbeuter()
{
	return Newsboat::configpaths::bridged::try_migrate_from_newsbeuter(*rs_object);
}

bool ConfigPaths::create_dirs() const
{
	return Newsboat::configpaths::bridged::create_dirs(*rs_object);
}

void ConfigPaths::set_cache_file(const std::string& new_cachefile)
{
	Newsboat::configpaths::bridged::set_cache_file(*rs_object, new_cachefile);
}

std::string ConfigPaths::url_file() const
{
	return std::string(Newsboat::configpaths::bridged::url_file(*rs_object));
}

std::string ConfigPaths::cache_file() const
{
	return std::string(Newsboat::configpaths::bridged::cache_file(*rs_object));
}

std::string ConfigPaths::config_file() const
{
	return std::string(Newsboat::configpaths::bridged::config_file(*rs_object));
}

std::string ConfigPaths::lock_file() const
{
	return std::string(Newsboat::configpaths::bridged::lock_file(*rs_object));
}

std::string ConfigPaths::queue_file() const
{
	return std::string(Newsboat::configpaths::bridged::queue_file(*rs_object));
}

std::string ConfigPaths::search_history_file() const
{
	return std::string(Newsboat::configpaths::bridged::search_history_file(*rs_object));
}

std::string ConfigPaths::cmdline_history_file() const
{
	return std::string(Newsboat::configpaths::bridged::cmdline_history_file(*rs_object));
}

} // namespace Newsboat

File: rust/libnewsboat-ffi/src/configpaths.rs
use crate::cliargsparser::CliArgsParser;
use libNewsboat::configpaths;
use std::path::Path;

// cxx doesn't allow to share types from other crates, so we have to wrap it
// cf. https://github.com/dtolnay/cxx/issues/496
pub struct ConfigPaths(configpaths::ConfigPaths);

#[cxx::bridge(namespace = "Newsboat::configpaths::bridged")]
mod bridged {
    #[namespace = "Newsboat::cliargsparser::bridged"]
    extern "C++" {
        include!("libNewsboat-ffi/src/cliargsparser.rs.h");
        type CliArgsParser = crate::cliargsparser::CliArgsParser;
    }

    extern "Rust" {
        type ConfigPaths;

        fn create() -> Box<ConfigPaths>;

        fn create_dirs(configpaths: &ConfigPaths) -> bool;

        fn initialized(configpaths: &ConfigPaths) -> bool;
        fn error_message(configpaths: &ConfigPaths) -> String;

        fn process_args(configpaths: &mut ConfigPaths, args: &CliArgsParser);

        fn try_migrate_from_newsbeuter(configpaths: &mut ConfigPaths) -> bool;

        fn url_file(configpaths: &ConfigPaths) -> String;
        fn cache_file(configpaths: &ConfigPaths) -> String;
        fn set_cache_file(configpaths: &mut ConfigPaths, path: &str);
        fn config_file(configpaths: &ConfigPaths) -> String;
        fn lock_file(configpaths: &ConfigPaths) -> String;
        fn queue_file(configpaths: &ConfigPaths) -> String;
        fn search_history_file(configpaths: &ConfigPaths) -> String;
        fn cmdline_history_file(configpaths: &ConfigPaths) -> String;
    }
}

fn create() -> Box<ConfigPaths> {
    Box::new(ConfigPaths(configpaths::ConfigPaths::new()))
}

fn create_dirs(configpaths: &ConfigPaths) -> bool {
    configpaths.0.create_dirs()
}

fn initialized(configpaths: &ConfigPaths) -> bool {
    configpaths.0.initialized()
}

fn error_message(configpaths: &ConfigPaths) -> String {
    configpaths.0.error_message().to_owned()
}

fn process_args(configpaths: &mut ConfigPaths, args: &CliArgsParser) {
    configpaths.0.process_args(&args.0);
}

fn try_migrate_from_newsbeuter(configpaths: &mut ConfigPaths) -> bool {
    configpaths.0.try_migrate_from_newsbeuter()
}

fn url_file(configpaths: &ConfigPaths) -> String {
    configpaths.0.url_file().to_string_lossy().into_owned()
}

fn cache_file(configpaths: &ConfigPaths) -> String {
    configpaths.0.cache_file().to_string_lossy().into_owned()
}

fn set_cache_file(configpaths: &mut ConfigPaths, path: &str) {
    configpaths.0.set_cache_file(Path::new(path).to_owned());
}

fn config_file(configpaths: &ConfigPaths) -> String {
    configpaths.0.config_file().to_string_lossy().into_owned()
}

fn lock_file(configpaths: &ConfigPaths) -> String {
    configpaths.0.lock_file().to_string_lossy().into_owned()
}

fn queue_file(configpaths: &ConfigPaths) -> String {
    configpaths.0.queue_file().to_string_lossy().into_owned()
}

fn search_history_file(configpaths: &ConfigPaths) -> String {
    configpaths
        .0
        .search_history_file()
        .to_string_lossy()
        .into_owned()
}

fn cmdline_history_file(configpaths: &ConfigPaths) -> String {
    configpaths
        .0
        .cmdline_history_file()
        .to_string_lossy()
        .into_owned()
}
